### PR TITLE
PR: Add translation package

### PR DESCRIPTION
**Pull request by goanpeca**
## References

Related to https://github.com/jupyterlab/jupyterlab/issues/3965

## Code changes

This adds a new translation and a translation-extension package using gettext to translate the JLab interface.

It provides a new ITranslator plugin which loads LanguageBundles that expose the gettext interface:
* gettext
* ngettext
* pgettext
* npgettext
etc...

## User-facing changes

A new menu `Language` has been added to the Settings menu

Screen Shot 2020-08-07 at 19 13 12
Screen Shot 2020-08-07 at 19 13 05

## Backwards-incompatible changes

The next PR that actually implements the usage will be backwards incompatibility, this one not yet.

### CODE REVIEW

File: packages/translation-extension/src/gettext.ts [hunk 1]
```diff
@@ -0,0 +1,671 @@
...
+  /**
+   * The domain of the translation, usually the normalized package name.
+   * Example: "jupyterlab", "jupyterlab_git"
+   */
+  domain?: string;
+
+  /**
+   * The delimiter to use when adding contextualized strings.
+   */
+  ctxt_delimiter?: string;
```
Comment (blink1073) [inline packages/translation-extension/src/gettext.ts:81]: Camel case?

File: packages/translation-extension/src/gettext.ts [hunk 2]
```diff
@@ -0,0 +1,671 @@
...
+  private _plural_forms: any;
+  private _dictionary: any;
+  private _locale: string;
+  private _domain: string;
+  private _ctxt_delimiter: string;
+  private _plural_funcs: any;
+  private _defaults: any;
+}
+
+let i18n = new Gettext({ domain: 'jupyterlab', locale: 'es-CO' });
```
Comment (blink1073) [inline packages/translation-extension/src/gettext.ts:622]: debug code

Comment (goanpeca) [reply packages/translation-extension/src/gettext.ts]: Fixed and added tests.

File: packages/translation-extension/src/gettext.ts (hunk 1 above)
Comment (goanpeca) [reply packages/translation-extension/src/gettext.ts]: Fixed

**Comment by goanpeca** [comment]
The js-apputils tests run locally and seem unrelated to changes on this PR.

File: packages/translation-extension/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+  ) => {
+    let setting = await settings.load(PLUGIN_ID);
+    let currentLocale: string = setting.get('locale').composite as string;
+    let languageData = await connector.fetch({ language: currentLocale });
+    let domainData: any = languageData?.data;
+    let englishBundle = new Gettext();
+    let translationBundle: Gettext;
+    let translationBundles: any = {};
+    let metadata: any;
+    console.log('Language data from server extension', domainData);
```
Comment (blink1073) [inline packages/translation-extension/src/index.ts:68]: console log

File: packages/translation-extension/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+          ],
+          1
+        );
+
+        let command: string;
+
+        // Get list of available locales
+        requestTranslationsAPI<any>('')
+          .then(data => {
+            console.log(data);
```
Comment (blink1073) [inline packages/translation-extension/src/index.ts:155]: console log

File: packages/translation-extension/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+                    buttons: [Dialog.cancelButton(), Dialog.okButton()]
+                  }).then(result => {
+                    if (result.button.accept) {
+                      setting
+                        .set('locale', locale)
+                        .then(() => {
+                          window.location.reload();
+                        })
+                        .catch(reason => {
+                          console.log(reason);
```
Comment (blink1073) [inline packages/translation-extension/src/index.ts:187]: Should be console.error

File: packages/translation/src/server.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+): Promise<T> {
+  // Make request to Jupyter API
+  const settings = ServerConnection.makeSettings();
+  const requestUrl = URLExt.join(
+    settings.baseUrl,
+    '/lab/api/translations/', // API Namespace
+    locale
+  );
+
+  console.log(requestUrl);
```
Comment (blink1073) [inline packages/translation/src/server.ts:27]: console log

File: packages/translation/src/server.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    throw new ServerConnection.NetworkError(error);
+  }
+
+  let data: any = await response.text();
+
+  if (data.length > 0) {
+    try {
+      data = JSON.parse(data);
+    } catch (error) {
+      console.log('Not a JSON response body.', response);
```
Comment (blink1073) [inline packages/translation/src/server.ts:42]: console.error

**Comment by blink1073** [comment]
Merge and iterate time!

**Pull request merged**